src/operator/filter.rs

```diff
@@ -72,30 +69,43 @@ where
     }
     fn clock_start(&mut self, _scope: Scope) {}
     fn clock_end(&mut self, _scope: Scope) {}
+    fn fixedpoint(&self) -> bool {
+        true
+    }
 }
 
-impl<K, V, CI, CO, F> UnaryOperator<CI, CO> for FilterKeys<K, V, CI, CO, F>
+impl<CI, CO, F> UnaryOperator<CI, CO> for FilterKeys<CI, CO, F>
 where
-    K: Clone + 'static,
-    V: Clone + 'static,
-    CI: IntoIterator<Item = (K, V)> + 'static,
-    for<'a> &'a CI: IntoIterator<Item = (&'a K, &'a V)>,
-    CO: FromIterator<(K, V)> + 'static,
-    F: Fn(&K) -> bool + 'static,
+    CI: BatchReader<Time = ()> + 'static,
+    CI::Key: Clone,
+    CI::Val: Clone,
+    CO: Batch<Key = CI::Key, Val = CI::Val, Time = (), R = CI::R> + 'static,
+    F: Fn(&CI::Key) -> bool + 'static,
 {
     fn eval(&mut self, i: &CI) -> CO {
-        i.into_iter()
-            .filter_map(|(k, v)| {
-                if (self.filter)(k) {
-                    Some((k.clone(), v.clone()))
-                } else {
-                    None
+        let mut cursor = i.cursor();
+
+        // We can use Builder because cursor yields ordered values.  This
+        // is a nice property of the filter operation.
+        let mut builder = CO::Builder::with_capacity((), i.len());
```

Comment:
does this cause waste if the filter throws out many elements?

Reply:
It does actually, since these buffers can make it all the way to the output batch.  This is probably ok, because the batch will either get freed at the end of the current clock tick or get added to the trace, where it will likely get merged with other batches soon, at which point the waste is gone.

I'll add a comment, and we'll keep an eye on this in profiling results.